### PR TITLE
ARROW-10680: [Rust] [DataFusion] Add partial support for TPC-H query 12

### PR DESCRIPTION
**Pull request by andygrove**
This PR adds partial support for TPC-H query 12 and fixes a couple of bugs that I found in the new JOIN logic.

### CODE REVIEW

**Comment by andygrove** [comment]
The query runs for a while and then fails with:

```
Running benchmarks with the following options: BenchmarkOpt { query: 12, debug: false, iterations: 1, concurrency: 2, batch_size: 4096, path: "/mnt/tpch/tbl-sf1/", file_format: "tbl", mem_table: false }
thread 'main' panicked at 'must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`', datafusion/src/physical_plan/hash_aggregate.rs:368:9
```

@jorgecarleitao @alamb I suggest that we merge this anyway since it does fix two join bugs and the benchmark is valid.

**Review by jorgecarleitao — APPROVED** [review]
Ehehe, now is when we start getting to the fun stuff of scheduling. I would place a "WARNING" or "EXPERIMENTAL" somewhere, just so that if someone is trying it in the wild they know that it is there.

File: rust/benchmarks/src/bin/tpch.rs [hunk 1]
```diff
@@ -89,6 +89,8 @@ enum TpchOpt {
     Convert(ConvertOpt),
 }
 
+const TABLES: &[&'static str] = &["lineitem", "orders"];
```
Comment (jorgecarleitao) [inline rust/benchmarks/src/bin/tpch.rs:92]:
```suggestion
const TABLES: &[&str] = &["lineitem", "orders"];
```

**Review by jorgecarleitao — COMMENTED** [review]
Helping fixing formatting :)

**Comment by andygrove** [comment]
@jorgecarleitao the error was caused by a tokio version mismatch between the crates

**Comment by andygrove** [comment]
The query works if I disable the `FilterPushDown` rule:

```
Running benchmarks with the following options: BenchmarkOpt { query: 12, debug: true, iterations: 1, concurrency: 24, batch_size: 4096, path: "/mnt/tpch/tbl-sf1/", file_format: "tbl", mem_table: false }
Logical plan:
Aggregate: groupBy=[[#l_shipmode]], aggr=[[SUM(Int32(1)) AS high_line_count, SUM(Int32(0)) AS low_line_count]]
  Join: l_orderkey = o_orderkey
    Filter: #l_receiptdate Lt Utf8("1995-01-01")
      Filter: #l_receiptdate GtEq Utf8("1994-01-01")
        Filter: #l_shipdate Lt #l_commitdate
          Filter: #l_commitdate Lt #l_receiptdate
            Filter: #l_shipmode Eq Utf8("MAIL") Or #l_shipmode Eq Utf8("SHIP")
              TableScan: lineitem projection=None
    TableScan: orders projection=None
Optimized logical plan:
Aggregate: groupBy=[[#l_shipmode]], aggr=[[SUM(Int32(1)) AS high_line_count, SUM(Int32(0)) AS low_line_count]]
  Join: l_orderkey = o_orderkey
    Filter: #l_receiptdate Lt Utf8("1995-01-01")
      Filter: #l_receiptdate GtEq Utf8("1994-01-01")
        Filter: #l_shipdate Lt #l_commitdate
          Filter: #l_commitdate Lt #l_receiptdate
            Filter: #l_shipmode Eq Utf8("MAIL") Or #l_shipmode Eq Utf8("SHIP")
              TableScan: lineitem projection=Some([0, 10, 11, 12, 14])
    TableScan: orders projection=Some([0])
+------------+-----------------+----------------+
| l_shipmode | high_line_count | low_line_count |
+------------+-----------------+----------------+
| MAIL       | 15526           | 0              |
| SHIP       | 15462           | 0              |
+------------+-----------------+----------------+
Query 12 iteration 0 took 13431 ms
```

**Comment by jorgecarleitao** [comment]
I will take a look at the filter pushdown.

**Comment by andygrove** [comment]
build failure seems unrelated:

```
LINK : fatal error LNK1318: Unexpected PDB error; FILE_SYSTEM (3) 'D:\a\arrow\arrow\rust\target\debug\deps\arrow-0d5ca67d15d3b3d6.pdb'
```

**Comment by rdettai** [comment]
@andygrove @jorgecarleitao I noticed you updated tokio to 0.3 in arrow-flight. You need to activate the `rt-multi-thread` for it to work! 😃 


And apart from that, don't you have problems running tonic (which uses tokio 0.2) with tokio 0.3 ?

`cd rust/arrow-flight; cargo run --example server`
-> 'there is no reactor running, must be called from the context of Tokio runtime'

**Comment by rdettai** [comment]
We can discuss downgrading back to tokio 0.2 in the causing PR https://github.com/apache/arrow/pull/8697 or open an issue

File: rust/benchmarks/src/bin/tpch.rs [hunk 2]
```diff
@@ -193,8 +143,105 @@ async fn benchmark(opt: BenchmarkOpt) -> Result<()> {
...
+                .table("lineitem")?
+                .filter(
+                    col("l_shipmode")
+                        .eq(lit("MAIL"))
+                        .or(col("l_shipmode").eq(lit("SHIP"))),
+                )?
+                .filter(col("l_commitdate").lt(col("l_receiptdate")))?
+                .filter(col("l_shipdate").lt(col("l_commitdate")))?
+                .filter(col("l_receiptdate").gt_eq(lit("1994-01-01")))?
+                // we do not support date functions yet, so faking the "+ interval '1' year" part
```
Comment (alamb) [inline rust/benchmarks/src/bin/tpch.rs:214]: there is `to_timestmp`: https://github.com/apache/arrow/blob/master/rust/datafusion/src/physical_plan/datetime_expressions.rs#L81 but we still need support to support intervals

**Review by alamb — COMMENTED** [review]
🎉

**Pull request closed without merging**